### DISCUSSION TITLE
Should Evennia implement markdown styling instead of ANSI flags?

### DISCUSSION THREAD

**Discussion by InspectorCaracal** [evennia/evennia#2746] [Evennia Engine Development]
I've been thinking a lot lately about the "ANSI-first" approach in formatting messages with Evennia. Particularly how it creates a lot of difficulties when converting to something browser-friendly, and working alongside MXP, and such.

In my opinion, the fact that Evennia supports building custom protocol portals and supports a webclient out of the box is one of the most valuable features in terms of infrastructure, but a good amount of that benefit is muted by the game messages being very telnet-first in design. I think something that better suits the top-level, protocol-agnostic design approach of the overall structure of the engine would make it easier to adapt and expand it without losing telnet compatability.

To that end, owllex introduced me to [the Rich library](https://pypi.org/project/rich/) which allows you to use markdown and pseudo-BBCode to style text, and then converts that to ANSI codes. There's plenty of libraries that will then convert markdown to HTML tags, as well, and having been working with the text2html parser lately, I believe the color flag and styling codes that Rich uses would be much more easily parsed into HTML spans than the current ANSI switches, making it overall easier to style your game for multiple end-protocols.

All that said... while I think it would be a really good idea, it would be a _very_ big change and make a major impact on backwards compatability with 0.9.5 so it's not something to just be stuck in there lightly.

What do y'all think?

**Comment by Griatch** [comment]
Yeah, this is an interesting notion. The plan (for Evennia 1.x) is to more thoroughly decouple the messages "as seen on the server" from what is presented over the wire. For the client we control (the webclient), we could even do the conversion client-side entirely. 

First thought has been to use the `|r` style locally only, but there is a point in using a more formal structure - markdown is so commonly prevalent that it may be a good candidate. Volund presented the Rich library in chat quite some time ago now, he also made some MXP additions to it, to make it work better in a MUD context. 

So yes, this is a good idea ... but it will be a lot of work to add; most likely we'll refactor the Session mechanism before this.

**Reply by InspectorCaracal** [reply]
Yeah, that's what I was thinking, I love the idea but it would be a huge amount of work to add.

I'm looking forward to that decoupling, though! I've been making a lot of notes re: that on my own, since I want to make my own webclient etc. without having to completely rewrite the whole infrastructure. There's a lot of stuff that can be done with the current infrastructure, still, and I've been working on the webclient parsing (as you've probably noticed xD) but being able to use full markdown formatting and have it stripped or converted as appropriate down the respective protocols would be *really* nice. Especially for picking up newcomers who aren't already familiar with traditional MUD formats.

What's the Session refactor plan look like? I haven't heard about it before this.

**Reply by Griatch** [reply]
> What's the Session refactor plan look like? I haven't heard about it before this.

The rough design is here: https://github.com/evennia/evennia/issues/2171 I already implemented it in a prototype, but it requires so many changes that it didn't feel suitable for a 0.9x -> 1.0 change (it just grew the scope too much).

**Comment by volundmush** [comment]
Rich is an awesome library and I fully back its use for solving any problems involving ANSI.

I was thinking of using it for my own projects but now I realize integrating it into Evennia may be simpler than I first thought.

https://github.com/volundmush/mudrich/

MudRich here has some hacks for enabling MXP to render within Rich and also adds more of the Python string API to the rich.text.Text object, among other things it does - like make Text, Tables, and Columns able to be PICKLED. There may need to be more work done but I could see pickling the Rich objects and sending it over AMP from Server to Portal and have the Rich Console objects on the Portal, deciding whether to export as ansi or html.
